api/src/main/java/com/velocitypowered/api/util/Identifiable.java

```diff
@@ -0,0 +1,11 @@
+package com.velocitypowered.api.util;
+
+import java.util.UUID;
+
+/**
+ * Represents an object that can be identified by its UUID
+ */
+public interface Identifiable {
```

Comment:
What purpose does this interface really have? I don't see any real reason for this.

Reply:
No reason yet in your implementation but it can be useful for some plugins to generalize some method related to these objects. In Velocity, the only affected classes are `Player` and `GameProfile` but as an example, a plugin might want to have a generic method to retrieve some data from its database according the id of the object passed as argument, whether it's a `Player`, a `GameProfile` or any custom object implementing `Identifiable`. Another example, it can be used with an `instanceof` to check whether an object has an unique id attached or not. A such interface can't hurt, but only help plugin developers. 😄 You will say, okay, so let them add their own interface but it would loose the interest to have a common a interface for every plugin and the API.

Reply:
I'm still not sold.

> a plugin might want to have a generic method to retrieve some data from its database according the id of the object passed as argument

There is already a standard object plugins can use to represent an UUID. It is called `java.util.UUID`. And getting the UUID is trivially done, given that there are only two objects in the Velocity API that even have a UUID:

```java
public static UUID getUuid(Object o) {
  if (o instanceof Player) return ((Player) o).getUniqueId();
  if (o instanceof GameProfile) return ((GameProfile) o).idAsUuid();
  throw new IllegalArgumentException("Provided object " + o + " doesn't have a UUID");
}
```

If a plugin needs a custom object for representing a user, it's probably going to implement a method similar to what I've provided for Velocity above.

There is no direct use of `Identifiable` in the Velocity API I can see providing convenience and I honestly don't think there is a use of this interface where a plain `java.util.UUID` would equally suffice.

Even more embarrassingly, let's assume you adjusted `ProxyServer#getPlayer()` to take an `Identifiable`. Does this code make sense?

```java
Player p = event.getPlayer();

// ... a lot of code ...
Player p1 = server.getPlayer(p); // but you already have the Player!
```

> You will say, okay, so let them add their own interface but it would loose the interest to have a common a interface for every plugin and the API.

But the common interface we could define here simply does not exist. An example of a common interface would be [PermissionFunction](https://jd.velocitypowered.com/com/velocitypowered/api/permission/PermissionFunction.html) - a common interface for plugins to define what permissions some subject has. It's useful - this is the foundation of the Velocity permisisons system. Where's the common interface and use for `Identifiable`? None, as far as I can tell.

Reply:
```java
Player p = event.getPlayer();

// ... a lot of code ...
Player p1 = server.getPlayer(p); // but you already have the Player!
```

I agree this code is almost useless I more thinking of an example like this one:
Imagine another plugin added a custom CommandSource also implementing Identifable.
Now, in my plugin, I'm implementing this command:

```java
  @Override
  public void execute(CommandSource source, String[] args) {
    if (source instanceof Identifiable) { // Can't be replaced without sacrificing compatibility with others plugins
      User user = User.of((Identifiable) source); // Can be replaced by User.ofId(((Identifiable) source).getUniqueId())
      source.sendMessage(TextComponent.of("Hello " + user.getDisplayName() + " !"));
    } else {
      source.sendMessage(TextComponent.of("Sorry, but this command is not available for a source of your type :(."));
    }
  }
```

Once again it's a detail. If you really don't want this interface, just tell me and I will remove it. To finish it makes things slightly more homogenous. I don't see any reason to call two identical methods (`Player#getUniqueId` and `GameProfile#idAsUuid`) with differents names.

Reply:
I'm still not sold on it. I'd prefer if `Identifiable` was removed.

Reply:
Done. I would have tried 😶